apk/apk.go

```diff
@@ -0,0 +1,424 @@
+	mw := io.MultiWriter(digest, w)
+	gw := gzip.NewWriter(mw)
+	cw := newWriterCounter(gw)
+	bw := bufio.NewWriterSize(cw, 4096)
+	tw := tar.NewWriter(bw)
+
+	err := builder(tw)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle the cut vs full tars
+	if err = bw.Flush(); err != nil {
+		return nil, err
+	}
+	if err = tw.Close(); err != nil {
+		return nil, err
+	}
+	if kind == tarFull {
+		if err = bw.Flush(); err != nil {
+			return nil, err
+		}
+	}
+
+	size := cw.Count()
+	alignedSize := (size + 511) & ^uint64(511)
+
+	increase := alignedSize - size
+	if increase > 0 {
+		b := make([]byte, increase)
+		_, err = cw.Write(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err = gw.Close(); err != nil {
+		return nil, err
+	}
+
+	return digest.Sum(nil), nil
+}
+
+func parseRsaPrivateKeyFromPemStr(privPEM string) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(privPEM))
+	if block == nil {
+		return nil, errors.New("failed to parse PEM block containing the key")
+	}
+
+	priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return priv, nil
+}
+
+/*func main() {
+	if err := runit("foo", "../alpine/user.rsa", "", os.Args[1]); err != nil {
+		log.Fatalln(err)
+	}
+}
+*/
+
+func runit(pathToFiles, pathToKey, workDir, target string) (err error) {
+	signatureTgz, err := os.Create(path.Join(workDir, "apk_signatures.tgz"))
+	if err != nil {
+		return err
+	}
+	defer signatureTgz.Close()
+
+	controlTgz, err := os.Create(path.Join(workDir, "apk_control.tgz"))
+	if err != nil {
+		return err
+	}
+	defer controlTgz.Close()
+
+	dataTgz, err := os.Create(path.Join(workDir, "apk_data.tgz"))
+	if err != nil {
+		return err
+	}
+	defer dataTgz.Close()
+
+	size := int64(0)
+	sizep := &size
+
+	// create the data tgz
+	dataDigest, err := createGzData(dataTgz, pathToFiles, sizep, size)
+	if err != nil {
+		return err
+	}
+
+	// create the control tgz
+	builderControl := createBuilderControl(size, dataDigest, err)
+	controlDigest, err := writeTgz(controlTgz, tarCut, builderControl, sha256.New())
+	if err != nil {
+		return err
+	}
+	fmt.Println("data sha1  :", hex.EncodeToString(controlDigest))
+
+	// pemBytes, err := ioutil.ReadFile("../alpine/user.rsa")
+	// pemBytes, err := ioutil.ReadFile("/home/appuser/.ssh/id_rsa")
+	pemBytes, err := ioutil.ReadFile(filepath.Clean(pathToKey))
+	if err != nil {
+		return err
+	}
+	priv, err := parseRsaPrivateKeyFromPemStr(string(pemBytes))
+	if err != nil {
+		return err
+	}
+	signed, err := priv.Sign(rand.Reader, controlDigest, crypto.SHA256)
+	if err != nil {
+		return err
+	}
+	fmt.Println("data sign  :", hex.EncodeToString(signed))
+
+	// create the signature tgz
+	builderSignature := createBuilderSignature(signed, err)
+	_, err = writeTgz(signatureTgz, tarCut, builderSignature, sha256.New())
+
+	// combine
+	return combineGzsToFile(target, signatureTgz, controlTgz, dataTgz)
```

Comment:
I'd either name it `combineTgzsTo()` or just `combineToApk()`

Reply:
> I'd either name it `combineTgzsTo()` or just `combineToApk()`

Done in 9d5bbd8